packages/flutter_tools/lib/src/base/features.dart

```diff
@@ -0,0 +1,245 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../version.dart';
+import 'config.dart';
+import 'context.dart';
+import 'platform.dart';
+
+/// The current [FeatureFlags] implementation.
+///
+/// If not injected, a default implementation is provided.
+FeatureFlags get featureFlags => context.get<FeatureFlags>() ?? const FeatureFlags();
+
+/// The interface used to determine if a particular [Feature] is enabled.
+///
+/// The rest of the tools code should use this class instead of looking up
+/// features directly. To faciliate rolls to google3 and other clients, all
+/// flags should be provided with a default implementation here. Clients that
+/// use this class should extent instead of implement, so that new flags are
+/// picked up automatically.
+class FeatureFlags {
+  const FeatureFlags();
+
+  /// Whether flutter web is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isWebEnabled => false;
+
+  /// Whether flutter desktop for macOS is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isMacOSEnabled => false;
+
+  /// Whether flutter desktop for Windows is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isWindowsEnabled => false;
+
+  /// Whether flutter desktop for linux is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isLinuxEnabled => false;
+}
+
+/// All current Flutter feature flags.
+const List<Feature> allFeatures = <Feature>[
+  flutterWebFeature,
+  flutterLinuxDesktopFeature,
+  flutterMacOSDesktopFeature,
+  flutterWindowsDesktopFeature,
+];
+
+/// The [Feature] for flutter web.
+const Feature flutterWebFeature = Feature(
+  name: 'Flutter Web',
+  configSetting: 'flutter-web',
+  environmentOverride: 'FLUTTER_WEB',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+  dev: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for macOS desktop.
+const Feature flutterMacOSDesktopFeature = Feature(
+  name: 'Flutter Desktop for macOS',
+  configSetting: 'flutter-macos-desktop',
+  environmentOverride: 'ENABLE_FLUTTER_DESKTOP',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for Linux desktop.
+const Feature flutterLinuxDesktopFeature = Feature(
+  name: 'Flutter Desktop for Linux',
+  configSetting: 'flutter-linux-desktop',
```

Comment:
nit: what about making the `configSetting` the same as the `environmentOverride`?

Reply:
We want to be able to turn on different desktop platforms separately, so they need to be separate settings, but we previously had `ENABLE_FLUTTER_DESKTOP` as an environment variable control so using that means that older third-party instructions about using desktop will continue to work.

Reply:
Got it. What about using `configSetting: 'enable-flutter-desktop'`, so we know that the equivalent env flag is `ENABLE_FLUTTER_DESKTOP`?

Reply:
But with the new approach we want to be able to turn on each desktop platform individually as they stabilize, rather than be stuck with the all-or-nothing of the original environment variable.

Reply:
I see.